l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol

```diff
@@ -0,0 +1,560 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 Aztec Labs.
+pragma solidity >=0.8.27;
+
+import {ISlasher} from "@aztec/core/interfaces/ISlasher.sol";
+import {IEmperor} from "@aztec/governance/interfaces/IEmpire.sol";
+import {IPayload} from "@aztec/governance/interfaces/IPayload.sol";
+import {SignatureLib, Signature} from "@aztec/shared/libraries/SignatureLib.sol";
+import {Slot, Timestamp, Epoch} from "@aztec/shared/libraries/TimeMath.sol";
+import {TimeLib} from "@aztec/core/libraries/TimeLib.sol";
+import {Errors} from "@aztec/core/libraries/Errors.sol";
+import {IValidatorSelection} from "@aztec/core/interfaces/IValidatorSelection.sol";
+import {SlashPayload} from "@aztec/periphery/SlashPayload.sol";
+
+type Round is uint256;
+
+/**
+ * @notice A SlashingProposer implementation based on consensus voting
+ */
+contract ConsensusSlashingProposer {
+  using SignatureLib for Signature;
+  using TimeLib for Slot;
+  using TimeLib for Epoch;
+
+  /**
+   * @notice Round data stored in circular storage
+   * TODO: Have a "Compressed" version of this struct to save gas
+   */
+  struct RoundData {
+    Round roundNumber; // The actual round number (for staleness check)
+    uint256 voteCount; // Number of votes collected so far
+    Slot lastVoteSlot; // Last slot for which a vote was received
+    bool executed; // Whether this round has been executed
+    // QUESTION: Each element of this array is defined as a dynamically-sized `bytes` array, but we know its size is
+    // `COMMITTEE_SIZE * ROUND_SIZE_IN_EPOCHS / 2` exactly. Can we use that to optimize how this is stored?
```

Comment:
Could we not make it fixed size [][] so it is not necessary to keep track of the size.